proposed/resource-limit.md

```diff
@@ -0,0 +1,843 @@
+
+We also plan on adding extension methods to simplify the common scenario where a count of 1 is used to represent 1 request.
+
+```c#
+    public static class ResourceLimiterExtensions
+    {
+        public static Resource Acquire(this ResourceLimiter limiter)
+        {
+            return limiter.Acquire(1);
+        }
+
+        public static ValueTask<Resource> AcquireAsync(this ResourceLimiter limiter, CancellationToken cancellationToken = default)
+        {
+            return limiter.AcquireAsync(1, cancellationToken);
+        }
+    }
+```
+
+#### Common usage patterns
+
+In an ASP.NET Core endpoint, usage of specific `ResourceLimiter`s will look like the following. Note that in these cases, the resource limiter is not known to be a rate or concurrency limiter and therefore the release mechanics must be followed.
+
+```c#
+ResourceLimiter limiter = new SomeResourceLimiter(options => ...)
+
+// Synchronous checks
+endpoints.MapGet("/acquire", async context =>
+{
+    // Check limiter using `Acquire` that should complete immediately
+    using var resource = limiter.Acquire();
+    // Resource was successfully obtained, the using block ensures
+    // that the resource is released upon processing completion.
+    if (resource.IsAcquired)
+    {
+        await context.Response.WriteAsync("Hello World!");
+    }
+    else
+    {
+        // Resource could not be obtained, send 429 response
+        context.Response.StatusCode = StatusCodes.Status429TooManyRequests;
+        return;
+    }
+}
+
+// Async checks
+endpoints.MapGet("/acquireAsync", async context =>
+{
+    // Check limiter using `AcquireAsync` which may complete immediately
+    // or wait until resources are available. Using block ensures that
+    // the resource is released upon processing completion.
+    using var resource = await limiter.AcquireAsync();
+    if (resource.IsAcquired)
+    {
+        await context.Response.WriteAsync("Hello World!");
+    }
+    else
+    {
+        // This exception may be thrown when hard cap is reached
+        context.Response.StatusCode = StatusCodes.Status429TooManyRequests;
+        return;
+    }
+}
+```
+
+In cases where it is known that the `ResourceLimiter` is for certain a rate limiter, for example in cases where the user of the limiter is also the instantiator of the limiter, the usage can be simplified to ignore the release mechanic:
+
+```c#
+ResourceLimiter rateLimiter = new SomeRateLimiter(options => ...)
+
+// Synchronous checks
+endpoints.MapGet("/acquire", context =>
+{
+    // Check limiter using `Acquire` that should complete immediately
+    if (rateLimiter.Acquire().IsAcquired)
```

Comment:
```suggestion
    using var resource = rateLimiter.Acquire();
    if (resource .IsAcquired)
```

Reply:
Here I'm explicitly illustrating when you know for sure it is a rate limiter, for example when you were the one to create the instance, you can skip the using pattern for even simpler usage.